contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+        _checkRole(resource, role, _msgSender());
+        _;
+    }
+
+    /**
+     * @dev Modifier that checks that sender has a specific role within the root resource. 
+     * Reverts with an {AccessControlUnauthorizedAccount} error including the required role.
+     */
+    modifier onlyRootRole(uint256 role) {
+        _checkRole(ROOT_RESOURCE, role, _msgSender());
+        _;
+    }
+
+    /**
+     * @dev Constructor.
+     *
+     * @param initialAdmin The address to grant the DEFAULT_ADMIN_ROLE to.
+     */
+    constructor(address initialAdmin) {
+        _grantRoles(ROOT_RESOURCE, DEFAULT_ADMIN_ROLE, initialAdmin);
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(EnhancedAccessControl).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted all roles in the root resource.
+     */
+    function hasRootRoles(uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return (_roles[ROOT_RESOURCE][account] & rolesBitmap) != 0;
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted all roles in `resource`.
+     */
+    function hasRoles(bytes32 resource, uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return ((_roles[resource][account] & rolesBitmap) != 0) || hasRootRoles(rolesBitmap, account);
+    }
+
+
+    /**
+     * @dev Returns the admin role that controls `role`. See {grantRole} and
+     * {revokeRole}.
+     *
+     * To change a role's admin, use {_setRoleAdmin}.
+     */
+    function getRoleAdmin(uint256 role) public view virtual returns (uint256) {
+        return _adminRoles[role];
+    }
+
+    /**
+     * @dev Returns the locked roles bitmap for a resource.
+     */
+    function getLockedRoles(bytes32 resource) public view virtual returns (uint256) {
+        return _lockedRoles[resource];
+    }
+
+    /**
+     * @dev Grants role to `account`.
+     *
+     * If `account` had not been already granted `role`, emits a {RoleGranted}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleGranted} event.
+     */
+    function grantRole(bytes32 resource, uint256 role, address account) public virtual canGrantRole(resource, role) returns (bool) {
+        return _grantRoles(resource, role, account);
+    }
+
+
+    /**
+     * @dev Revokes `role` from `account`.
+     *
+     * If `account` had been granted `role`, emits a {RoleRevoked} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleRevoked} event.
+     */
+    function revokeRole(bytes32 resource, uint256 role, address account) public virtual canGrantRole(resource, role) returns (bool) {
```

Comment:
```suggestion
    function revokeRoles(bytes32 resource, uint256 role, address account) public virtual canGrantRole(resource, role) returns (bool) {
```

Reply:
This isn't as straightforward since we need to check that the caller has the admin role for the given role - and we don't want to be looping through and checking the caller is an admin for every role passed in.

The alternative is to be able to set an admin for a bitmap of roles, but then that semantically doesn't make any sense.

Hence why granting/revoking one role at a time makes the most sense.